File: impls/java/src/mal/env.java
package mal;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

import mal.types.MalList;
import mal.types.MalSymbol;
import mal.types.MalType;

public class env {

	public env outer;
	Map<MalSymbol, MalType> data;
	static types malTypes = new types();

	public env(env outer, MalList binds, MalList exprs) {
		this.outer = outer;
		this.data = new HashMap<>();
		for (int i = 0; binds != null && exprs != null && i < Math.min(binds.items.size(), exprs.items.size()); i++) {
			MalSymbol key = (MalSymbol) binds.items.get(i);
			MalType value = exprs.items.get(i);
			this.set(key, value);
		}
	}

	public void set(MalSymbol key, MalType value) {
		data.put(key, value);
	}

	public env find(MalSymbol key) {
		if (data.containsKey(key)) {
			return this;
		} else {
			if (outer == null) {
				throw malTypes.new RepException(String.format("'%s' not found", key.name));
			} else {
				return this.outer.find(key);
			}
		}
	}

	public MalType get(MalSymbol key) {
		env env = this.find(key);
		if (env == null) {
			throw malTypes.new RepException(String.format("'%s' not found", key));
		} else {
			return env.data.get(key);
		}
	}

	public String toString() {
		return String.join(", ",
				data.keySet().stream().map(k -> k.toString()).collect(Collectors.toList()));
	}
}

File: impls/clojure/src/mal/env.clj
(ns mal.env
  (:refer-clojure :exclude [find get set]))

(declare set)

(defn make-env
  ([outer] (make-env outer [] []))
  ([outer binds exprs]
   (let [newenv {:outer outer :data (atom {})}]
     (doall
      (map (partial set newenv) binds exprs))
     newenv)))

(defn set [this key val]
  (when-let [data (:data this)]
    (swap! data assoc key val)))

(defn find [this key]
  (if (clojure.core/get @(:data this) key)
    this
    (if-let [outer (:outer this)]
      (find outer key)
      (throw (ex-info "" {:cause (str "Env containing " key " not found")})))))

(defn get [this key]
  (if (some #{key} (keys @(:data this)))
    (clojure.core/get @(:data this) key)
    (if-let [outer (:outer this)]
      (get outer key)
      (throw (ex-info "" {:cause (format "'%s' not found" key)})))))

